### PR TITLE
Add mod action to verify file integrity

### PR DESCRIPTION
**Pull request by asarium**
This completely replaces the previous CheckTask which caused some issues
if the file hashes were not found in the hash cache. This has been
replaced by a dedicated action which can be triggered by the user when
desired.

The action currently only validates the files and displays an error
dialog if any invalid files have been found. This could be extended to
automatically redownload the corrupted or missing files but that has not
been included here.
f0299a5

### CODE REVIEW

File: knossos/web.py [hunk 1]
```diff
@@ -1214,6 +1212,14 @@ def parseIniMod(self, path):
         mod.load(path)
         return json.dumps(mod.get())
 
+    @QtCore.Slot(str)
+    def verifyModIntegrity(self, mid):
+        mod = self._get_mod(mid)
```
Comment (ngld) [inline knossos/web.py]: The version number is missing. This way it'll always check the newest installed version instead of the intended one.

File: knossos/tasks.py [hunk 2]
```diff
@@ -269,33 +194,58 @@ def init2(self):
...
+                else:
+                    subpath = os.path.relpath(path, modpath)
 
-            for item in files:
-                name = os.path.join(subpath, item).replace('\\', '/')
-                if name not in fnames and not name.startswith(('__k_plibs', 'knossos.')):
-                    loose.append(name)
+                for item in files:
+                    name = os.path.join(subpath, item).replace('\\', '/')
+                    if name not in fnames and not name.startswith(('__k_plibs', 'knossos.')):
```
Comment (ngld) [inline knossos/tasks.py]: You're comparing paths (`name`) with filenames here (`fnames`) this WILL fail for every mod that uses subfolders.
Why did you change the loop that builds the `fnames` list? It should work correctly without the change.

File: knossos/web.py [hunk 3]
```diff
@@ -141,7 +141,6 @@ def query(self, mid, spec=None):
     @QtCore.Slot()
     def fetchModlist(self):
         tasks.run_task(tasks.FetchTask())
-        tasks.run_task(tasks.CheckTask())
```
Comment (ngld) [inline knossos/web.py:144]: This should launch `LoadLocalModsTask` to pick up changes to local files.

File: knossos/web.py [hunk 4]
```diff
@@ -395,7 +394,6 @@ def setBasePath(self, path):
         else:
             center.settings['base_path'] = os.path.abspath(path)
             center.save_settings()
-            tasks.run_task(tasks.CheckTask())
```
Comment (ngld) [inline knossos/web.py:398]: After changing the base path we need to rebuild the list of installed mods. You **have to** run `LoadLocalModsTask` here.

File: html/templates/kn-mod-home.vue [hunk 5]
```diff
@@ -95,6 +99,7 @@ export default {
                         <button @click="uploadLog">Upload Debug Log</button>
                         <button v-if="mod.id !== 'FS2'" @click="install">Modify</button>
                         <button v-if="mod.id !== 'FS2' && !mod.dev_mode" @click="uninstallMod">Uninstall</button>
+                        <button v-if="!mod.dev_mode" @click="verifyIntegrity">Verify file integrity</button>
```
Comment (ngld) [inline html/templates/kn-mod-home.vue]: This should probably also exclude FS2 since we don't include checksums for FS2.

**Review by ngld — CHANGES_REQUESTED** [review]

File: html/templates/kn-mod-home.vue (hunk 5 above)
Comment (asarium) [reply html/templates/kn-mod-home.vue]: Ok, I thought FS2 supported the checksums.

File: knossos/tasks.py (hunk 2 above)
Comment (asarium) [reply knossos/tasks.py]: The reason I changed the `fname` building was that `name` was just the filename here for some reason while the filename from the package structure had a leading `./`.

I'll see what I can do about fixing this for mods with subfolders.

File: knossos/web.py (hunk 3 above)
Comment (asarium) [reply knossos/web.py:144]: Fixed.

File: knossos/web.py (hunk 4 above)
Comment (asarium) [reply knossos/web.py:398]: Fixed.

File: knossos/web.py (hunk 1 above)
Comment (asarium) [reply knossos/web.py]: Fixed and I also added the action to the details page.

File: knossos/tasks.py (hunk 2 above)
Comment (ngld) [reply knossos/tasks.py]: I think the filename is generated with `os.path.relpath` during upload. That might be the reason.

The easiest solution would probably be stripping `./` if it is found when creating the `fnames` list.
```python
if info['filename'].startswith('./'):
    fnames.add(info['filename'][2:])
else:
    fnames.add(info['filename'])
```

Comment (asarium) [reply knossos/tasks.py]: I think I found a better solution. I changed the code so that it only works on absolute paths which solved the `./` problem I had earlier. I also tested it with the FSPort MediaVPs which has files in subfolders and that also worked well.
